quickwit/quickwit-ingest/src/lib.rs

```diff
@@ -125,6 +125,49 @@ impl CommitType {
     }
 }
 
+struct TimedMutexGuard<T> {
+    guard: T,
+    acquired_at: std::time::Instant,
+    purpose: [&'static str; 2],
```

Comment:
one of the sad things with relying on dynamic labels like this, is that you end up doing one hashmap lookup very time you need to access the counter (here once per lock).

It is probably not catastrophic in this case, but here is the alternative.

have a histogram: directly on the `TimedMutexGuard`

```
histogram: &'static Histogram,
```

You can then cache the Histogram counter in different ways. In ingest v2 you can see the pattern done here https://github.com/quickwit-oss/quickwit/blob/main/quickwit/quickwit-ingest/src/ingest_v2/metrics.rs#L59-L74.

Reply:
i like this pattern of statically defining things, however it's a kind of a pain, and a large typo hazard to implement. Having some sort of macro to derive "Build a static (histogram|gauge|counter) vec from a dynamic vec" would be nice, and would likely make us use this more